src/mlpack/core/optimizers/sgd/test_function.cpp

```diff
@@ -58,3 +91,35 @@ void SGDTestFunction::Gradient(const arma::mat& coordinates,
       break;
   }
 }
+
+
+void SGDTestFunction::Gradient(const arma::mat& coordinates,
+                               const size_t begin,
+                               const size_t batchSize,
+                               arma::mat& gradient) const
+{
+  gradient.zeros(3);
+
+  for (size_t i = begin; i < begin + batchSize; ++i)
+  {
+    switch (visitationOrder(i))
+    {
+      case 0:
+        if (coordinates[0] >= 0)
+          gradient[0] += std::exp(-coordinates[0]);
+        else
+          gradient[0] += -std::exp(coordinates[0]);
+        break;
+
+      case 1:
+        gradient[1] += 2 * coordinates[1];
+        break;
+
+      case 2:
+        gradient[2] += 4 * std::pow(coordinates[2], 3) + 6 * coordinates[2];
+        break;
+    }
+  }
+
+  gradient /= batchSize;
```

Comment:
We should discuss this, do we divide the gradient inside the function or do we divide inside the optimize function. The same applies for the Evaluation function.

Reply:
What I thought was, when we use the gradient() in optimize(), we will be optimizing point wise. So, the gradient value should be that of the point and not of the batch itself. Also, I took the average as it serves the purpose of saving computation, which is the main reason for using batch processing.

Reply:
I get your point, I'm not sure someone likes to see the accumulated gradients, so I think taking the average makes sense. In that context I think it makes sense to also take the average over the loss. @Rajiv2605 @rcurtin What do you think?

Reply:
@zoq Yes, I also thought of the same thing.

Reply:
@zoq Hey, so are we averaging the loss?

Reply:
Sorry, I thought I had responded to this, I guess not.  I think it would be better to avoid taking the average, so that `Gradient(coordinates, 0, gradient, NumFunctions())` returns the exact same result as `Gradient(coordinates, gradient)`.